### PR TITLE
Introduce scheme

### PR DESCRIPTION
**Pull request by aalekseyev**
- Introduce a composable type `Scheme.t` for expressing lazily generated subtree rules.
- `Rules.t` representation is ugly and temporary. See #2089 where we turn it into a passive data structure instead.

### CODE REVIEW

File: src/file_binding.ml [hunk 1]
```diff
@@ -37,6 +37,15 @@ module Unexpanded = struct
 
   let expand_src t ~dir ~f = Path.relative dir (f t.src)
 
+  let destination_relative_to_install_path t ~f =
+    let dst = Option.map ~f t.dst in
+    match Install.Entry.adjust_dst_on_windows
+            ~src_basename:(fun () -> Filename.basename (f t.src))
+            ~dst
```
Comment (diml) [inline src/file_binding.ml]: Discussed with @aalekseyev in person: currently this imposes more restrictions on what variables can be used on Windows only, which is not great

File: bin/common.ml [hunk 2]
```diff
@@ -19,6 +19,7 @@ type t =
   { debug_dep_path        : bool
   ; debug_findlib         : bool
   ; debug_backtraces      : bool
+  ; debug_extra_backtraces      : bool
```
Comment (diml) [inline bin/common.ml]: This new feature doesn't seem to be used in this PR. Could you extract it in a separate PR? I also feel like it needs a convincing use case.

Comment (aalekseyev) [reply bin/common.ml]: Right. Sorry, I wanted to use this twice so far, but both use cases eventually disappeared. I'll keep it in a branch for now and wait until another use case pops up. 

The intended use case is to store backtraces on allocation of certain object (e.g. Memo.lazy_) to know where it came from.

File: src/dir_set.ml [hunk 3]
```diff
@@ -0,0 +1,196 @@
+open! Stdune
+
+module T = struct
+
+  type t = {
```
Comment (diml) [inline src/dir_set.ml]: Why do we have two records BTW? I'm also wondering if it be worth representing `Empty` and `Universal` as dedicated constructors in `t`

Comment (aalekseyev) [reply src/dir_set.ml]: Two records were just easier for me to think about because `children` can be thought of as a function [string -> t] and I expected some operations to work just on `children`.

Special representation of `Empty` and `Universal` is indeed not necessary, but I expected that to be a worthwhile optimization, given that we're testing things for whether they are Empty or Universal pretty often. They also have special status of being disallowed in certain contexts (I disallow the "useless" entries in the `exceptions` map).
Do you think we should remove the special representation?

Comment (diml) [reply src/dir_set.ml]: Do you mean add them back? It seems better to me to have them yes. Most operations match on it, and even for the one that don't it looks like it would still make them more efficient.

Comment (aalekseyev) [reply src/dir_set.ml]: I pushed this change.

File: src/rules.mli [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+(** [rule] is a function that produces some build system rules
+    such as ([Build_system.add_rule]) in a known directory. *)
+type rule = unit -> unit
+
+type t = private rule Path.Build.Map.t
+
+val to_map : t -> rule Path.Build.Map.t
+
+(* [Path] must be in build directory *)
+val file_rule : rule:(Path.t * rule) -> unit
```
Comment (diml) [inline src/rules.mli]: Should we use `Path.Build.t` here?

Comment (aalekseyev) [reply src/rules.mli]: Yes, as soon as `build_system.ml` is switched (this will be done in a follow-up feature).

Comment (aalekseyev) [reply src/rules.mli]: In fact this occurrence of `Path.t` is just removed later (because the `Path.t` information is duplicated in the rule itself):

```
https://github.com/ocaml/dune/pull/2089/files#diff-9cb90475c596eb47885cc09b7f1af8ce
```

Comment (diml) [reply src/rules.mli]: Ack

File: src/scheme.mli [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    | Empty
+    | Union of 'rules t * 'rules t
+    | Approximation of Dir_set.t * 'rules t
+    | Finite of 'rules Path.Build.Map.t
+    | Thunk of (unit -> 'rules t)
+end
+
+module type S = Scheme_intf.S with module Gen := Gen
+
+module Make(Directory_rules : sig
```
Comment (diml) [inline src/scheme.mli]: This API feels a bit odd to me. I.e. I don't really know how I would call `Directory_rules`; it is a module that can only produce empty values. Looking at the implementation, we only need `union` in `evaluate` and `empty` in `rules_here`. What do you think about only exposing the polymorphic API, passing `union` to `evaluate` and replacing `'rules` by `'rules option` in `rules_here`?

Comment (aalekseyev) [reply src/scheme.mli]: I like that. I'm making the change.

**Comment by diml** [comment]
BTW, would it be hard to extract 0a9c83a3261d23d99ea641053bab9aa4689bcb12 into a separate PR? It looks independant from this PR.

**Comment by aalekseyev** [comment]
@diml, I don't think it's independent.
Without that change, to use scheme we need to call `add_rule` for every directory we have rules for. The point of Scheme is to make that process demand-driven.

I could move it together with the rest of install_rules changes easily enough though. That would leave scheme unused in this feature.

**Comment by rgrinberg** [comment]
I will give this a review today. Sorry for the delay. Splitting this PR would be highly appreciated and make review easier for me.

**Comment by aalekseyev** [comment]
Moved the install rules changes to #2130

File: src/rules.mli [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  val empty : t
+  val union : t -> t -> t
+end
+
+type t = private Dir_rules.t Path.Build.Map.t
+
+val to_map : t -> Dir_rules.t Path.Build.Map.t
+
+(* [Path] must be in build directory *)
+val file_rule : rule:(Path.t * Dir_rules.t) -> unit
```
Comment (rgrinberg) [inline src/rules.mli]: Why not make the argument `Path.Build.t`?

File: src/rules.ml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+module T = struct
+  type t = Dir_rules.t Path.Build.Map.t
+
+  let empty = Path.Build.Map.empty
+
+  let union_map a b ~f =
+    Path.Build.Map.union a b ~f:(fun _key a b -> Some (f a b))
+
+  let union =
+    union_map ~f:(fun rule1 rule2 -> fun () -> rule1 (); rule2 ())
```
Comment (rgrinberg) [inline src/rules.ml]: How about using `Dir_rules.union` here? Perhaps it would be nicer to make `Dir_rules.t` abstract btw.

Comment (aalekseyev) [reply src/rules.ml]: Yeah, `Dir_rules.union` didn't exist when this line was written.

This is getting rewritten in #2089 anyway so I wouldn't worry too much.

File: src/rules.mli (hunk 6 above)
Comment (aalekseyev) [reply src/rules.mli]: This is being changed (the argument is removed altogether) in #2089 and fine-tuning it here will just increase merge conflicts, so I'd like to merge this as-is.

File: src/dir_set.ml [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+      if (x.here    && not y.here   ) ||
+         (x.default && not y.default) then
+        not_subset ();
+      ignore
+        (merge_exceptions x y ~default:false ~f:loop : t String.Map.t);
+      Empty
+  in
+  fun x ~of_ ->
+    match loop x of_ with
+    | (_ : t) -> true
```
Comment (rgrinberg) [inline src/dir_set.ml]: Why do we return `t` here and not `unit` for example?

File: src/scheme.ml [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+let evaluate ~union_rules ~env =
+  let rec loop = function
+    | Empty -> Evaluated.empty
+    | Union (x, y) ->
+      Evaluated.union ~union_rules (loop x) (loop y)
+    | Approximation (paths, rules) ->
+      if
+        not (Dir_set.is_subset paths ~of_:env)
+        && not (Dir_set.is_subset (Dir_set.negate paths) ~of_:env)
```
Comment (rgrinberg) [inline src/scheme.ml]: What is this `env` parameter for? It seems to always be `Universal`, so these `is_subset` calls will always be true. Do we expect it to vary down the line?

File: src/scheme.mli [hunk 10]
```diff
@@ -0,0 +1,20 @@
+(** A collection of rules for one or multiple directories. *)
+
+open! Stdune
+
+type 'rules t =
+  | Empty
+  | Union of 'rules t * 'rules t
+  | Approximation of Dir_set.t * 'rules t
+  | Finite of 'rules Path.Build.Map.t
+  | Thunk of (unit -> 'rules t)
```
Comment (rgrinberg) [inline src/scheme.mli]: I would add a comment to each of these constructors.

File: src/dir_set.ml (hunk 8 above)
Comment (diml) [reply src/dir_set.ml]: It was just so that I could write `merge... ~f:loop` which wants `f`  to return a `t`.

File: src/scheme.ml (hunk 9 above)
Comment (aalekseyev) [reply src/scheme.ml]: Ah, that's a bug. Thanks for catching. I pushed a fix.

The purpose of the parameter is to catch errors when rules are generated for directories outside of `env`. Speaking of which, that's not checked in the `Finite` branch at all.

I'm making a fix for that too.

**Comment by aalekseyev** [comment]
@rgrinberg, the utility is having fewer PRs in flight at a time :-) and having github understand what the diff of the follow-up PRs is.
